design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+**Environment-agnostic Stacks**
+
+When a stack is defined, users can specify `env` (account and/or region).
+
+If `account` and/or `region` use the pseudo references `Aws.ACCOUNT_ID` and `Aws.REGION`, respectively, the stack is called "environment-agnostic". Certain features in the CDK, like VPC lookups for example, are not supported for environment-agnostic stacks since the specific account/region is required during synthesis.
+
+The proposal described in [PR#4131](https://github.com/aws/aws-cdk/pull/4131) suggests that environment-agnostics stacks cannot be deployed using the CDK. However, it also proposes that the default behavior for `env` will be to use ("inherit") the CLI configured environment when a stack is deployed through `cdk deploy`.
+
+This means that the only way to produce environment-agnostic templates will be to explicitly indicate it when a stack is defined.
+
+Since the specific account and region are required when resolving asset consumption and publishing locations, the current plan is for the default asset store implementation to **fail if assets are used from environment-agnostic stacks**. Again, bear in mind that the current behavior (where the default is environment-agnostic stacks) is going to be changed.
+
+## Bootstrapping
+
+The CDK already has a dedicated tool for bootstrapping environments called **`cdk bootstrap`**. An environment is bootstrapped once, and from that point, it is possible to deploy CDK apps into this environment.
+
+Environment bootstrapping doesn't have to be performed by the development team, and does not require deep knowledge of the application structure, besides the set of accounts and regions into which the app needs to be deployed.
+
+The current implementation only provisions an S3 bucket, but in order to be able to continuously deploy CDK stacks that use asses, we will need the following resources:
+
+For publishing:
+
+* **S3 Bucket**: for file asset and CloudFormation templates
+* **ECR Repository**: for docker image assets
+* **Publishing Role**: IAM role trusted by the deployment account, and allows publishing to the S3 bucket and the ECR repository.
+
+For deployment:
+
+* **CloudFormation Role**: IAM role which allows the CloudFormation service principal to deploy stacks into the environment (this role usually has administrative privileges).
+* **Deployment Role**: IAM role which allows anyone from the deployment account to create, update and describe CloudFormation change sets and s"pass" the CloudFormation role (`iam:PassRole`).
+
+To accommodate these requirements we will make the following changes to how `cdk bootstrap` works:
+
+1. Extend the bootstrap stack to include these resources.
+2. Use explicit convention-based physical names for all resources.
+3. Allow specifying a list of *trusted accounts* from which the CloudFormation service principal can deploy to this account.
+4. Allow specifying a list of accounts from which principals can deploy to this account from the CLI.
```

Comment:
I don't understand what this point means. Can you clarify?

Reply:
Cleaned up